Add deploy/zyxel_gs1900.sh

coderjoe opened the pull request:
Add support for deploying to the Zyxel GS1900 line of switches as long as those switches are running at least firmware V2.80.

Tested on a Zyxel GS1900-8 and GS1900-24E

Resolves #5042

coderjoe commented:
I'm now using this version locally for my switches. Please let me know if there is anything you'd like me to do to improve the PR. No rush. :)

coderjoe commented:
A bug has occurred on one of my servers with a different version of sed. Working on a fix now.

coderjoe commented:
Most recent commit should have resolved the issue I found.

Neilpang commented on deploy/zyxel_gs1900.sh:
```diff
@@ -0,0 +1,423 @@
+  # to return a consistent body return - so we cannot inspect the result of this
+  # upload to determine success. We will need to re-query the certificates page
+  # and compare the validity dates to try and identify if they have changed.
+  upload_response_html=$($_CURL \
```
only for `curl` here?  how about `wget` ?
I would suggest you not use `curl` here,  please use `_post()` method instead.

coderjoe replied:
I originally followed the example in `synology_dsm.sh` and constructed the multpart form request by hand. This did not work in either `_post()` or `wget`. Both of those mangled the binary certificate data.

Specifically some of the shell-unsafe bytes and null bytes seemed to be getting mangled.

I am willing to try again, but I did not have success originally.

coderjoe replied:
I will try, but I'm pretty sure the server is expecting unencoded binary data for the certificate file in the multi part form upload.

coderjoe replied:
Thank you for your suggestion. My computer has broken and I'm now waiting on replacement parts.

Once I'm back up and running I will be trying what you suggest.

coderjoe replied:
I have tried your suggestion but I cannot make it work.
I will am updating the PR to draft, and updating the branch to see if you have any other ideas to avoid the `curl` requirement.

coderjoe commented on deploy/zyxel_gs1900.sh:
```diff
@@ -275,26 +275,41 @@ _zyxel_gs1900_upload_certificate() {
+    printf -- "%s\r\n" "${upload_post_boundary}"
+    printf "Content-Disposition: form-data; name=\"http_file\"; filename=\"cert.pfx\"\r\n"
+    printf "Content-Type: application/octet-stream\r\n\r\n"
+    cat "${temp_pkcs12}"
```
Note: the file located at `temp_pkcs12` contains null `\0` bytes.
As such it cannot be saved to a variable in many shells.

coderjoe replied:
Resolved by saving to a file and using --data-binary for curl and --post-file or --body-file for wget.

coderjoe commented on deploy/zyxel_gs1900.sh:
```diff
@@ -275,26 +275,41 @@ _zyxel_gs1900_upload_certificate() {
-    --form "sysSubmit=Import" \
-    "${_zyxel_switch_base_uri}/cgi-bin/httpuploadcert.cgi")
+  export _H2="Content-type: multipart/form-data boundary=${upload_post_boundary}"
+  _post "$(cat "${upload_post_request}")" "${_zyxel_switch_base_uri}/cgi-bin/httpuploadcert.cgi"
```
@Neilpang  this seems to be my problem. The body must be passed as a variable to `_post` but I cannot put null bytes `\0` in a variable in all shells.

This results in the `\0`s being stripped from the body and makes the body invalid.

I cannot url-encode this because the server requires the raw octet byte stream, not an encoded value.

`curl` supports accepting a raw file via the "var=@filename" format.
`wget` does not seem to support accepting raw file input.

Do you have any other ideas that I may be able to try?

coderjoe commented on deploy/zyxel_gs1900.sh:
```diff
@@ -275,26 +275,41 @@ _zyxel_gs1900_upload_certificate() {
-    -X "POST" \
-    -H "$_H1" \
-    --form "XSSID=${form_xss_value}" \
-    --form "http_file=@${temp_pkcs12}" \
```
Note the @ sign in this --form.
Many --form makes it a multipart form
The @ sign makes it read the file for the value of the form data and allows me to have a file with null bytes `\0` in them.

This is what I cannot reproduce.

coderjoe reviewed (commented):
These are the differences between the two implementations and why I was using `curl` rather than `_post`. I have not been able to get `_post` to work.

coderjoe commented:
I've updated for the shellcheck and shfmt linting in the previous run, have removed the defunct curl errors (now that curl is not a requirement) and have squashed the commit to clean up history a bit.

coderjoe commented:
I've verified that the updated commits continue to work in all of my deployed environments.
Thank you very much @Neilpang for your help/guidance on wget support.

I don't think I would have figured it out without your suggestion. :)

Neilpang commented on deploy/zyxel_gs1900.sh:
```diff
@@ -0,0 +1,425 @@
+  # requires a password be present. Then upload that certificate.
+  temp_cert_password=$(head /dev/urandom | tr -dc 'A-Za-z0-9' | head -c 64)
+  temp_cert_password="password"
+  _secure_debug2 "temp_cert_password" "$temp_cert_password"
```
temp_cert_password  always equals to "password" ?

coderjoe replied:
Ugh... rookie mistake... I thought I reverted that. Left over from testing binary cert upload issues ... 🤦‍♀️

coderjoe replied:
Fixed in latest, as well as removing some unnecessary stdout from the upload that isn't used by the switch.

coderjoe commented:
I have now been using this in production for a few months. @Neilpang please let me know if there is anything else you would like me to do - or if there's anything I can do to help get this included. :)

Thanks again for your time and help!

coderjoe commented:
Rebased onto the latest `dev`.
Please let me know what else needs to be done for a review.

Neilpang commented:
why do you need to "ReadBodyFromFile"?

coderjoe commented:
> why do you need to "ReadBodyFromFile"?

Great question @Neilpang. I responded in the gitter.im chat while the repo was restricted, but I'll repeat it here just in case it wasn't seen.

I need to read from file because this switch expects a binary exact, not encoded, upload of the certificate. The uploaded certificate contains many \0 NULL bytes which cannot be piped within the shell script without being truncated.

Making `wget` or `curl` read the certificate file directly is the only way I have been able to get the full file uploaded in the manner the switch expects without the file becoming truncated by the null bytes.

I hope this helps clarify.

dark-m0de commented:
Hi, I would really appreciate if we could add support for Zyxel devices. I'm using a XMG1915 switch, as well as NWA50AX PRO and NWA130BE access points. My assumption is that with a few changes, those devices could be supported as well.
I'd be happy to support the development for this.

coderjoe commented:
Hello @dark-m0de , I appreciate your interest. I don't think in its current form my PR will be accepted. I think my modifications to the _post method in acme.sh are not of sufficient quality and I'm currently investigating alternatives.

Some notes:

- Web interfaces between different series of Zyxel switches tend to be different enough that they will likely require separate integrations. At least that has been my past experience. 
- I'm entirely unfamiliar with their AP products so I cannot provide any recommendation, but much like the XGS series I would be very hesistant to support them in this script.
- I have an open [feature request on the Zyxel community](https://community.zyxel.com/en/discussion/comment/73153#Comment_73153) to suppor other methods of upload other than just web based upload of binary PKCS12 certificates. That conversation is ongoing at this time and if you'd like to leave a polite "me too" that would be welcome.

For now I think I'm going to put this PR back into draft until I can come up with some alternative methods to get the certificates into the GS1900 series switchs.

dark-m0de commented:
Thank you @coderjoe, I just prepared my answer for the zyxel community. But somehow I'm not able to post it yet.

Another thought: Do you think it would help if zyxel adds support for uploading certificates to their CLI (e.g. ssh based)? This would add ssh as a dependency to acme in such scenarios. But it would definitely ease the process to manage certificates in a scripted way.

coderjoe commented:
@Neilpang  I've updated this PR to the latest `dev` and removed all depependence on changes to the acme.sh base functionality. If you manage to find the time I'd appreciate another review.

Thank you in advance. :)

coderjoe commented:
Updated to correct the failures in the prior commit's ShellCheck and shfmt errors.

Neilpang commented:
add your usage here:
https://github.com/acmesh-official/acme.sh/wiki/deployhooks

coderjoe commented:
Documentation added to the wiki:
https://github.com/acmesh-official/acme.sh/wiki/deployhooks#36-deploying-to-zyxel-gs1900-series-switches

Neilpang commented:
please create a bug to accept reports:
https://github.com/acmesh-official/acme.sh/wiki/DNS-API-Dev-Guide#9-please-create-a-new-issue-for-future-bugs

coderjoe commented:
Bug report issue #6370 created and subscribed. :+1:
